### ISSUE TITLE
Include :: 'index out of range' exception thrown in some cases of multi-include queries (need to run both queries in the repro)

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFramework#1478]
Repro:

```

    public class Level1
    {
        public int Id { get; set; }
        public string Name { get; set; }

        public Level2 OneToOne_Required_PK { get; set; } 
        public Level2 OneToOne_Optional_PK { get; set; } 

        public Level2 OneToOne_Optional_FK { get; set; } 

        public ICollection<Level2> OneToMany_Optional { get; set; } 
        public ICollection<Level1> OneToMany_Optional_Self { get; set; } 
        public Level1 OneToMany_Optional_Self_Inverse { get; set; } 
    }

    public class Level2
    {
        public int Id { get; set; }
        public string Name { get; set; }

        public int Level1_Required_Id { get; set; }
        public int? Level1_Optional_Id { get; set; }

        public Level1 OneToOne_Required_PK_Inverse { get; set; }
        public Level1 OneToOne_Optional_PK_Inverse { get; set; }
        public Level1 OneToOne_Required_FK_Inverse { get; set; }
        public Level1 OneToOne_Optional_FK_Inverse { get; set; }

        public Level1 OneToMany_Optional_Inverse { get; set; } 
    }

    public class MyContext : DbContext
    {
        public DbSet<Level1> LevelOne { get; set; }
        public DbSet<Level2> LevelTwo { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            modelBuilder.Entity<Level1>().Property(e => e.Id).GenerateValueOnAdd(false);
            modelBuilder.Entity<Level2>().Property(e => e.Id).GenerateValueOnAdd(false);

            modelBuilder.Entity<Level1>().HasOne(e => e.OneToOne_Required_PK).WithOne(e => e.OneToOne_Required_PK_Inverse).ReferencedKey<Level1>(e => e.Id).Required(true); 
            modelBuilder.Entity<Level1>().HasOne(e => e.OneToOne_Optional_PK).WithOne(e => e.OneToOne_Optional_PK_Inverse).ReferencedKey<Level1>(e => e.Id).Required(false);
            modelBuilder.Entity<Level1>().HasOne(e => e.OneToOne_Optional_FK).WithOne(e => e.OneToOne_Optional_FK_Inverse).ForeignKey<Level2>(e => e.Level1_Optional_Id).Required(false);
            modelBuilder.Entity<Level1>().HasMany(e => e.OneToMany_Optional).WithOne(e => e.OneToMany_Optional_Inverse).Required(false);
            modelBuilder.Entity<Level1>().HasMany(e => e.OneToMany_Optional_Self).WithOne(e => e.OneToMany_Optional_Self_Inverse).Required(false);

            modelBuilder.Entity<Level2>().Property(e => e.Id).GenerateValueOnAdd(false);
            modelBuilder.Entity<Level2>().Property(e => e.Id).GenerateValueOnAdd(false);
        }

        protected override void OnConfiguring(DbContextOptions options)
        {
            options.UseSqlServer(new SqlConnectionStringBuilder { DataSource = ".", InitialCatalog = "Repro1478", MultipleActiveResultSets = true, IntegratedSecurity = true }.ToString());
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();

                var l1_01 = new Level1 { Id = 1, Name = "L1 01", };
                var l1_02 = new Level1 { Id = 2, Name = "L1 02", };
                var l1_03 = new Level1 { Id = 3, Name = "L1 03", };
                var l1_04 = new Level1 { Id = 4, Name = "L1 04", };

                var l2_01 = new Level2 { Id = 1, Name = "L2 01", };
                var l2_02 = new Level2 { Id = 2, Name = "L2 02", };
                var l2_03 = new Level2 { Id = 3, Name = "L2 03", };
                var l2_04 = new Level2 { Id = 4, Name = "L2 04", };

                var l1s = new[] { l1_01, l1_02, l1_03, l1_04 }.ToArray();
                var l2s = new[] { l2_01, l2_02, l2_03, l2_04 }.ToArray();


                ctx.LevelOne.Add(l1s);
                ctx.LevelTwo.Add(l2s);

                ctx.SaveChanges();

                l1s[0].OneToOne_Optional_PK = l2s[0];
                l1s[2].OneToOne_Optional_PK = l2s[2];

                l1s[1].OneToOne_Optional_FK = l2s[2];
                l1s[3].OneToOne_Optional_FK = l2s[0];

                l1s[0].OneToMany_Optional = new List<Level2> { l2s[1], l2s[3], };

                l1s[1].OneToMany_Optional_Self = new List<Level1> { l1s[0] };
                l1s[3].OneToMany_Optional_Self = new List<Level1> { l1s[2] };

                ctx.SaveChanges();
            }

            using (var ctx = new MyContext())
            {
                var q1 = ctx.LevelOne
                    .Include(e => e.OneToMany_Optional)
                    .ThenInclude(e => e.OneToMany_Optional_Inverse)
                    .ThenInclude(e => e.OneToMany_Optional_Self_Inverse)
                    .ThenInclude(e => e.OneToOne_Optional_FK).ToList();

                var q2 = ctx.LevelOne
                    .Include(e => e.OneToMany_Optional)
                    .ThenInclude(e => e.OneToMany_Optional_Inverse)
                    .ThenInclude(e => e.OneToMany_Optional_Self_Inverse)
                    .ThenInclude(e => e.OneToOne_Optional_PK).ToList();
            }
        }
    }

```

exception:

Unhandled Exception: System.ArgumentOutOfRangeException: Index was out of range.
 Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argu
ment, ExceptionResource resource)
   at System.Collections.Generic.List`1.get_Item(Int32 index)
   at Microsoft.Data.Entity.Query.QueryBuffer.GetTargetEntity(IEntityType target
EntityType, IValueReader valueReader, ICollection`1 bufferedEntities) in D:\k\En
tityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:line 432
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyLi
st`1 navigationPath, IReadOnlyList`1 relatedValueReaders, Int32 currentNavigatio
nIndex) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:li
ne 190
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyLi
st`1 navigationPath, IReadOnlyList`1 relatedValueReaders, Int32 currentNavigatio
nIndex) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:li
ne 192
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyLi
st`1 navigationPath, IReadOnlyList`1 relatedValueReaders, Int32 currentNavigatio
nIndex) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:li
ne 192
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyLi
st`1 navigationPath, IReadOnlyList`1 relatedValueReaders) in D:\k\EntityFramewor
k\src\EntityFramework.Core\Query\QueryBuffer.cs:line 154
   at Microsoft.Data.Entity.Relational.Query.QueryMethodProvider.<>c__DisplayCla
ss11_0`1.<_Include>b__2(T qss) in D:\k\EntityFramework\src\EntityFramework.Relat
ional\Query\QueryMethodProvider.cs:line 92
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at Microsoft.Data.Entity.Relational.Utilities.EnumerableExtensions.Finally[TS
ource](IEnumerable`1 source, Action finallyAction) in D:\k\EntityFramework\src\E
ntityFramework.Relational\Utilities\EnumerableExtensions.cs:line 19
   at Microsoft.Data.Entity.Relational.Query.QueryMethodProvider._Include[T](Rel
ationalQueryContext queryContext, IEnumerable`1 innerResults, IQuerySource query
Source, IReadOnlyList`1 navigationPath, IReadOnlyList`1 includeRelatedValuesStra
tegyFactories) in D:\k\EntityFramework\src\EntityFramework.Relational\Query\Quer
yMethodProvider.cs:line 101
   at lambda_method(Closure , QuerySourceScope )
   at System.Linq.Enumerable.<SelectManyIterator>d__8`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at Microsoft.Data.Entity.Query.EntityQueryExecutor.EnumerableExceptionInterce
ptor`1.EnumeratorExceptionInterceptor.MoveNext() in D:\k\EntityFramework\src\Ent
ityFramework.Core\Query\EntityQueryExecutor.cs:line 179
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Repro1478.Program.Main(String[] args) in D:\Projects\Repro1478\Repro1478\P
rogram.cs:line 118

**Comment by maumar** [comment]
most likely same root cause as #1487

**Comment by maumar** [comment]
fixed in 617ad0aaa964949e4a620ef5a73fa4bbe2f12374
